transport/tcp/tcp.go

```diff
@@ -125,7 +126,16 @@ func (d *dialer) Dial() (_ transport.Pipe, err error) {
 		return nil, err
 	}
 
-	return transport.NewConnPipe(conn, d.proto, d.opts)
+	p, err := transport.NewConnPipe(conn, d.proto, d.opts)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+	if err = d.handshaker.Start(p); err != nil {
+		conn.Close()
+		return nil, err
+	}
+	return d.handshaker.Wait()
```

Comment:
The combination of `Start` + `Wait` suspends the `Dial()` call until the `worker` has completed the handshake.
With tls+tcp we [ran into problems where the initial Dial() handshake never completed](https://github.com/nanomsg/mangos/issues/96#issuecomment-482187557), hence the Dial() code would still hang if the handshake never completed.

Reply:
Yes, that's correct.  The assumption here is that Dial() in the client is synchronous.  Most of the time this is appropriate.  For the cases where it isn't, the caller can set the OptionDialAsynch option, which basically causes this to run in a go routine.

Reply:
Oh, I was talking from the v1 perspective, which used DialAsynch. In our case, it [got stuck](https://github.com/nanomsg/mangos/issues/96#issuecomment-482187557) and was never able to report an error, causing the application to hang.